### PR TITLE
Value objects

### PR DESCRIPTION
**Pull request by Actimel**
Solve #98, #99

### CODE REVIEW

**Comment by Actimel** [comment]
It is not only about defining value object for sync actions. It's data are also used for generating data for manifest file. Many extractors may contain different fields which does not allow us to unify that value object.

Table with different fields:
# Action getTables
- `Y` - yes
- `N` - no
- `o` - may not be defined

## Tables[]
| Keys | Common | Mysql | PgSQL | Oracle | DB2 | MSSQL |
|---|---|---|---|---|---|---|
|`columns[]`| Y | Y | Y | Y | Y | Y |
|`name`| Y | Y | Y | Y | Y | Y |
|`sanitizedName`| Y | Y | N | N | N | N |
|`schema`| Y | Y | Y/`null` | Y | Y | Y |
|`type`| Y | Y | Y/`null` | N | Y | Y |
|`rowCount`| Y | Y | N | o | N | N |
|`tablespaceName`| N | N | N | Y | N | N |
|`catalog`| N | N | N | N | N | Y |
|`owner`| N | N | N | Y | N | N |
|`autoIncrement`| o | o | N | N | N | N |
|`timestampUpdateColumn`| o | o | N | N | N | N |

## Columns[]

| Keys | Common | Mysql | PgSQL | Oracle | DB2 | MSSQL |
|---|---|---|---|---|---|---|
|`name`| Y | Y | Y | Y | Y | Y |
|`sanitizedName`| Y | Y | Y | Y | N | Y |
|`type`| Y | Y | Y | Y | Y | Y |
|`primaryKey`| Y | Y | Y | Y | Y | Y |
|`length`| Y | Y | Y | Y | Y | Y |
|`nullable`| Y | Y | Y | Y | Y | Y |
|`default`| Y | Y | Y | N | Y | N |
|`ordinalPosition`| Y | Y | Y | Y | Y | Y |
|`constraintName`| o | o | N | N | N | N |
|`foreignKeyRefSchema`| o | o | N | N | N | o |
|`foreignKeyRefTable`| o | o | N | o | o | o |
|`foreignKeyRef`| N | N | N | o | o | N |
|`foreignKeyTable`| N | N | N | o | N | N |
|`primaryKeyName`| N | N | N | o | N | o |
|`uniqueKey`| N | N | N | Y | o | o |
|`uniqueKeyName`| N | N | N | o | N | o |
|`indexed`| N | N | N | N | o | N |
|`foreignKeyRefColumn`| o | o | N | N | N | o |
|`foreignKey`| N | N | N | N | N | o |
|`foreignKeyName`| N | N | N | N | N | o |
|`extra`| o | o | N | N | N | N |
|`autoIncrement`| o | o | N | N | N | o |
|`timestampUpdateColumn`| o | o | N | N | N | N |
|`checkConstraint`| N | N | N | N | N | o |
|`checkClause`| N | N | N | N | N | o |

**Comment by pivnicek** [comment]
@Actimel Those tables are a great help, do you mind if I add Redshift and Snowflake there?

**Comment by Actimel** [comment]
@pivnicek Sure you can, just please separete them (e.g. with some flag) that SF and RS are not working on ex-db-common

**Comment by pivnicek** [comment]
@Actimel ?  They are indeed on db-common

**Comment by Actimel** [comment]
@pivnicek I was checking extractors from https://github.com/keboola/db-extractor-common/blob/master/README.md#extractors-using-dbc
I see it now. But both of them using version 5.2 - that's a bit scary.

Anyway we should add those repositories into readme too.
https://github.com/keboola/db-extractor-snowflake
https://github.com/keboola/db-extractor-redshift

**Comment by pivnicek** [comment]
@Actimel Yes, snowflake has 7 PRs I want to merge (which bring it up to 9.2.1 etc)  you can approve them if you like:)

**Comment by pivnicek** [comment]
Ah, I didn't realise they weren't in the readme, I've added them now, thanks!

File: tests/Old/CommonExtractorTest.php [hunk 1]
```diff
@@ -537,9 +538,6 @@ public function testMetadataManifest(): void
             $metadataList[$metadata['key']] = $metadata['value'];
         }
 
-        $this->assertEquals(2, $metadataList['KBC.rowCount']);
```
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/Old/CommonExtractorTest.php]: Toto je jen ukazkovy extraktor, ktery testuje sice realny beh, ale neni to v nem potreba (rozumnej puvodni common byl 1:1 jako mysql extraktor), toho jsem se chtel zbavit, aby v common bylo jen nezbytne minimum a pridal to az v db-ex-mysql.

File: src/DatabaseMetadata/Table.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        return [
+            'name' => $this->name,
+            'sanitizedName' => $this->sanitizedName,
+            'schema' => $this->schema,
+            'type' => $this->type,
+            'columns' => $this->getColumns(),
+        ];
+    }
+
+    public function addColumn(int $position, Column $columnMetadata): void
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: Urcite dobra pripominka! Ale hure proveditelna, museli bychom k tomu zrefaktorovat i `getTables()`, coz bych nechal jako upravu do budoucna. Takto mi to prijde good enough, i kdyz tu pozici, by to mohlo prijimat jen `Column` a to poradi si vytahnout z tama. V pripade, ze chces prepsat existujici, tak zarvat.

File: src/DatabaseMetadata/Table.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    /** @var Column[] */
+    private $columns = [];
+
+    public function __construct(
+        string $name,
+        string $schema,
+        string $type
+    ) {
+        $this->name = $name;
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: Jak pises nesmi to byt empty string, ale IMHO k tomu nikdy nenastane, kdyz se to bere z information schema (pripadne z jine tabulky v zavislosti na konkretni databazi).
Nicmene urcite to stoji za zvazeni, pro pripad, ze by to nekdo chtel zneuzit jinak.

File: tests/CommonExtractor.php [hunk 4]
```diff
@@ -14,7 +14,10 @@
 use Keboola\DbExtractorCommon\Exception\ApplicationException;
```
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/CommonExtractor.php]: Smazal jsem ho v poslednim commitu, byl to pozustatek po rebasu. K nicemu uz neni potreba. Vidis na tom neco spatneho?

File: src/DatabaseMetadata/Table.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/DatabaseMetadata/Table.php]: Není potřeba to extra refaktorovat. Jen prostě metody ->set* nad Column a Table přesuneš do Builderu. Nebo nějak prohodíš generování columns a tables... 

Mimo to, getTables je abstraktní. A rozhodně bych nekazil API jen kvuli tomu, že je to "nějak" napsané v tom CommonExtractoru. 

Jakmile to jednou necháme takhle, tak už to pak bude navždycky mutable, protože se na to přepíšou všechny extractory. Tohle je přesně ta breaking změna, kterou chceme udělat teď IMHO.

File: tests/CommonExtractor.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/CommonExtractor.php]: No, že nebyl smazaný. Ale asi jsem jen neměl refreshnuto, sorry.

File: tests/CommonExtractor.php [hunk 5]
```diff
@@ -14,7 +14,10 @@
...
 use Keboola\DbExtractorCommon\Exception\DeadConnectionException;
 use Keboola\DbExtractorCommon\BaseExtractor;
+use Keboola\DbExtractorCommon\IncrementalFetchingSettings;
 use Keboola\DbExtractorCommon\RetryProxy;
+use Keboola\DbExtractorCommon\Tests\Old\CommonExtractorColumnMetadata;
```
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/CommonExtractor.php]: Jo, presne nad necim takovym jsem premyslel, Common mi prijde totalne zbytecny a hlavne neudrzitelny (uz ted se hodne zmenilo i v ex-db-mysql, ze ktereho vychazi). Ale zalezi jakou cestou se chceme vydat. Podminka je, ze to musi byt postavene nad db-ex-common, vzhledem k tomu, ze se prepisuji abstraktni metody pro dany usecase.

File: src/DatabaseMetadata/Table.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/DatabaseMetadata/Table.php]: Aspoň nějaká základní validace by byla dobrá. Ne proto, že to někdo bude zneužívat, ale protože se upíše a prohodí name a schema někde třeba...

File: tests/Functional/DatadirTest.php [hunk 6]
```diff
@@ -119,7 +119,6 @@ public function testActionGetTables(): void
                     'sanitizedName' => 'table1',
                     'schema' => 'testdb',
                     'type' => 'BASE TABLE',
-                    'rowCount' => '0',
```
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/Functional/DatadirTest.php]: Na poradi nezalezi, ale v tomto poradi to ted leze z `jsonSerialize()`.

File: tests/Functional/DatadirTest.php [hunk 7]
```diff
@@ -129,7 +128,7 @@ public function testActionGetTables(): void
                             'length' => '128',
                             'nullable' => false,
                             'default' => null,
-                            'ordinalPosition' => '1',
+                            'ordinalPosition' => 1,
```
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/Functional/DatadirTest.php]: KBC bych dot tohoto nezatahoval. Ten string lezl primo z extraktoru, ale nikdy nebyl pretypovan i kdyz tam nemuze byt nic jine nez `int`

File: src/BaseExtractor.php [hunk 8]
```diff
@@ -266,8 +268,9 @@ protected static function getDataTypeMetadata(array $column): array
         return $dataType->toMetadata();
     }
 
-    public static function getColumnMetadata(array $column): array
+    public static function getColumnMetadata(Column $column): array
     {
+        $column = JsonHelper::decode((string) json_encode($column));
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/BaseExtractor.php]: `json_encode()` jsem prehledl. Prislo mi ale zbytecne implementovat `toArray()`, kdyz to muzeme prohnat pres toto, i kdyz by to asi vypadalo lepe.

File: tests/Functional/DatadirTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/Functional/DatadirTest.php]: Myslel jsem to tak, že "doteď tam byl string a teď je tam integer" - tj. to se může propsat JSONem až někam do metadat ve storage. A kdyby někde v kódu storage bylo třeba `setOrdinalPosition(string $position)`, tak to na tom vyhučí. Tak se jen pro jistotu ptám, jestli to fakt nevadí.

File: tests/Functional/DatadirTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/Functional/DatadirTest.php]: OK, proč ne asi...

File: src/BaseExtractor.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/BaseExtractor.php]: Tak ono by stačilo i jen zavolat $tableDetails->jsonSerialize() když už ti nevadí, že je to takové nehezké... 

Tak ono by možná fungovalo i jen (array) :D 

Update: Skoro, ale ne :) https://3v4l.org/1W3rT

File: tests/CommonExtractor.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/CommonExtractor.php]: No a to je otázka, jestli je správně. Co když by bylo lepší mít `Db\Extractor\Adapter\MySQLAdapter`, který by implementovat `DbExtractorAdapterInterface` to by byl základní a minimální skill, který má každý adapter mít. Více méně by to mohlo být `fetchAllRows($tableName): iterable`. Pak by byly další interfacy, kterými by tomu adaptéru přibývaly schopnosti např `CanLoadIncrementallyInterface::loadIncremental($tableName, $lastValue): iterable` a `CanLoadIncrementallyInterface::isValidIncrementalConfiguration($column, $lastValue, ...)` nebo `CanDisplayTablesInterface::getTables()`. A v DbExtractoru by zbylo jen to, co všechno se má kde udělat. Ale už ne jak přesně. Třeba na loadování CSV by zas klidně mohl být `BasicCsvWriter::write(iterable $data)`, který by šel přes dependency injection přehodit za `SlicedCsvWriter implements DbExtractorCsvWriter` nebo za něco úplně jiného, pokud by to ten konkrétní extraktor potřeboval. 

Composition over inheritance!

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php [hunk 9]
```diff
@@ -46,9 +46,7 @@ protected function getDbNode(): ArrayNodeDefinition
 
         // @formatter:off
         $node
-            ->ignoreExtraKeys(false)
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: Jedine co zmizelo je klic `driver` a `password` (ktery jsme se shodli, ze z tama vyhodime, nikdo by ho uz pouzivat nemel). Proto bych tu ten ignore nenechaval a jen zpet dal ten `driver` (viz par radku nize).

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php [hunk 10]
```diff
@@ -188,7 +185,6 @@ protected function getTablesNode(): ArrayNodeDefinition
                     ->integerNode('retries')
                         ->min(0)
                     ->end()
-                    ->booleanNode('advancedMode')->end()
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: Je to pozustatek, ktery tu musime nechat, protoze je mozne, ze to nekdo stale ma v konfiguraci. `advancedMode = true` v pripade, ze pouzivas na export vlastni query.

File: src/DatabaseMetadata/Column.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\DbExtractorCommon\DatabaseMetadata;
+
+class Column implements \JsonSerializable
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Column.php]: Pridane

File: src/DatabaseMetadata/Column.php [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+{
+    /** @var string */
+    private $name;
+
+    /** @var string|null */
+    private $sanitizedName;
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Column.php]: Opraveno

File: src/DatabaseMetadata/Column.php [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+        return $result;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function getSanitizedName(): ?string
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Column.php]: Opravene

File: src/DatabaseMetadata/Table.php [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $this->columns[$position] = $columnMetadata;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function getSanitizedName(): ?string
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: Opraveno

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php [hunk 15]
```diff
@@ -63,7 +61,6 @@ protected function getDbNode(): ArrayNodeDefinition
                     ->isRequired()
                 ->end()
                 ->append($this->getSshNode());
-            //->end();
```
[the remark this answers is not in the document]
Comment (Actimel) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: Pridal jsem to tam.

File: tests/CommonExtractor.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/CommonExtractor.php]: Presne takovou predstavu jsem mel hned na zacatku, nez jsem se do toho pustil. Nicmene v tehdejsim stavu to nebylo moc proveditelne (ted si bohuzel nevzpomenu na cem jsem tolik narazil). V aktualnim stavu, kdyz se nebudes tolik ohlizet na to, ze to dedi spolecny Common, tak to vychazi v koncovych extraktorech temer stejne. Staci naimplementovat abstraktni metody a v pripade, ze chces neco delat jinak, tak pretizis rodicovske.

Nemyslim si, ze je to dedeni tady na skodu.

File: tests/Functional/DatadirTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (Actimel) [reply tests/Functional/DatadirTest.php]: IMHO se toto zpracovava z kbc-ui nejakym JS, ktery toto resit nebude. Ale urcite to zkontroluju jeste

File: src/DatabaseMetadata/Table.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: Asi mas pravdu. Je to vpodstate prevzate ze starsi verze, kdy se nekdy (i kdyz si nedokazu predstavit stavy, kdy to nastane) hodnota jak `name` nebo `schema` nastavuje prave ne empty string

[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: https://github.com/keboola/db-extractor-common/blob/7f9df3686b5ef2cdd0e2764762ba8d637b47abfc/src/Keboola/DbExtractor/Extractor/Common.php#L211-L213

[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/DatabaseMetadata/Table.php]: Ale tohle je situace, kdy to taháš třeba z Oraclu a "schema" tam prostě není. Jenže to není tenhle případ - tohle je abstrakce nad tou DB, ne? Jen proto, že se tomu tam třeba říká "tablegroup" tak se to nebude měnit - interně to pro nás bude furt schéma. Tj. tohle považuju furt za nevyřešené.

File: src/IncrementalFetchingSettings.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    /** @var string */
+    private $type;
+
+    /** @var int|null */
+    private $limit;
+
+    public function __construct(string $column, string $type)
+    {
+        $this->column = $column;
+        $this->type = $type;
```
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/IncrementalFetchingSettings.php]: Nevyřešené.

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php [hunk 17]
```diff
@@ -46,7 +46,6 @@ protected function getDbNode(): ArrayNodeDefinition
 
         // @formatter:off
```
Comment (tomasfejfar) [inline src/Configuration/Definition/BaseExtractorConfigDefinition.php]: Chybí ještě `/** @noinspection NullPointerExceptionInspection */`

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: No právě, že ho tam musíme nechat a tady je odebraný, proto se ptám.

[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: dtto ConfigRowDefinition.php

File: src/DatabaseMetadata/Column.php [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+    public function __construct(
+        string $name,
+        string $type,
+        bool $primaryKey,
+        ?string $length,
+        bool $nullable,
+        ?string $default,
+        int $ordinalPosition
+    ) {
+        $this->name = $name;
```
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply src/DatabaseMetadata/Column.php]: nevyřešeno

File: src/DatabaseMetadata/Table.php [hunk 19]
```diff
@@ -4,124 +4,78 @@
...
 
-    /** @var string|null */
+    /** @var string */
     private $schema;
 
-    /** @var string|null */
+    /** @var string */
     private $type;
 
-    /** @var int|null */
```
Comment (tomasfejfar) [inline src/DatabaseMetadata/Table.php]: tyhle jsou vyhozené protože je to mysql specific?

File: src/DatabaseMetadata/ToArrayInterface.php [hunk 20]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\DbExtractorCommon\DatabaseMetadata;
+
+interface ToArrayInterface
```
Comment (tomasfejfar) [inline src/DatabaseMetadata/ToArrayInterface.php]: Nevim, jestli mi to nepřijde zbytečný. Samo o sobě mi to nepřijde použitelný. Typicky nebudeš hintovat na ToArrayInterface - spíš na ten value object...

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (Actimel) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: https://github.com/keboola/db-extractor-common/pull/94#discussion_r226453949
Tady bych ho smazal a pridal az do konkretniho extraktoru (asi jen mysql)

File: src/DatabaseMetadata/Table.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: To jestli se nekdo upise a prohodi `name` a `schema`, tak te validace neohlida.
Jakou by sis predstavoval minimalni validaci pro toto, krome prazdneho stringu a pripadne dovolenych znaku, ktere se ale muzou pro ruzne databaze lisit?

[the remark this answers is not in the document]
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: FYI: https://github.com/keboola/db-extractor-common/pull/103#issuecomment-437165201 Tady je tabulka co ve kterych extraktorech se pouziva za pole.

File: src/DatabaseMetadata/Table.php (hunk 19 above)
Comment (Actimel) [reply src/DatabaseMetadata/Table.php]: Presne tak

File: src/Configuration/Definition/BaseExtractorConfigDefinition.php (hunk 17 above)
Comment (Actimel) [reply src/Configuration/Definition/BaseExtractorConfigDefinition.php]: IMO to neni potreba ho tam zbytecne davat. Ty chyby jsou vyignorovane v `phpstan.neon`

**Pull request closed without merging**